Address remaining nullable reference type issues (#1515)
dotnet/csharpstandard: standard/types.md

The ***default null state*** of an expression is determined by its type, and the state of the annotations flag when it is declared:

> - The default state of the `null` literal (§6.4.5.7) is maybe null.

Reviewer: @RexJaeschke Please file an issue (assign it to me) for expressions without types, referring to this PR.

- The default state of a default value expression (§12.8.21) is not null when the type is a non-nullable value type, otherwise maybe null.

> - For all other expressions

Reviewer 2: Is a colon needed?
```suggestion
- For all other expressions:
```

Author: I defer to @RexJaeschke  I think so.

Reviewer 3: Yes!

>   - The default null state of a nullable reference type is:

Author: ```suggestion
  - The default null state of an expression whose type is a nullable reference type is:
```

    - Maybe null when its declaration is in text where the annotations flag is enabled.
    - Not null when its declaration is in text where the annotations flag is disabled.

>   - The default null state of a non-nullable reference type is not null.

Author: ```suggestion
  - The default null state of an expression whose type is a non-nullable reference type is not null.
```

>
> *end example*

> A compiler may issue a warning when the result of a lifted conversion operator (§10.6.2) is the `null` reference and that value is assigned to a non-nullable reference type.

Reviewer 2: What about something like this, because the compiler doesn't know it _is_ the null reference, it just knows it is _maybe_ the null reference?
```suggestion
A compiler may issue a warning when the result of a lifted conversion operator (§10.6.2) is a reference type and is converted to a non-nullable reference type, since the result of a lifted conversion operator is maybe-null.
```
Or alternatively we say "the default null state of a lifted conversion operator that returns a reference type is maybe-null" and skip talking about the "assigned to a non-nullable reference type" part of this, since the warning-on-assignment part of this is the same for all maybe-null expressions no matter how they came to be maybe-null.
